### DISCUSSION TITLE
A lot of errors to publish. How to do it properly?

### DISCUSSION THREAD

**Discussion by LuisAlfredo92** [unoplatform/uno#11276] [Q&A] [answered]
Hello everybody, I'm requesting help because I've had a lot of problems to publish a Windows app (and a few with Android). I've used every possible method but nothing works. I'm sorry if I use some incorrect terms but I use another language on Visual Studio.

Just in case, it's a WinUI program, I have the Mobile, Skia.GTK, Windows and shared projects

# Using "Publish" menu
## Creating my own profile
Using the Publish option right-clicking on my Windows project, I've created a lot of publish profiles with different combinations but none has worked:
| Configuration | Implementation mode | Target runtime environment | Error message |
| ------------ | ------------ | ------------ | ------------ |
| Release - Any CPU | Framework dependent | Portable | It compiles but when I try to open the .exe it opens a window and closes immediately |
|  Release - Any CPU | Framework dependent | win10-x64 | The same as previous |
|  Release - x64 | Framework dependent | win10-x64 | The same as previous* |
|  Release - x64 | Framework independent | win10-x64 | The same as previous* |
|  Release - x86 | Framework independent | win10-x86 | Metadata file wasn't found 'PathToApp\App\App\App\bin\Release\net7.0-windows10.0.18362\win10-x86\_2FoxyAuth.dll' |
|  Release - x86 | Framework dependent | win10-x86 | The same as previous |

*Those with asterisk gave a different error (The same x86 gave), but I compiled the App.dll shared (and I'm not sure if I did it correctly) and then they started compiling successfully but the program didn't work

## Using ClickOnce profile
So I used ClickOne to make a profile with this configuration:

- **Configuration:** Release | Any CPU
- **Implementation mode:** Framework dependent
- **Target runtime environment:** Portable
And then I get this error: `The item "C:\Users\92lui\.nuget\packages\uno.winui\4.7.37\lib\net5.0-windows10.0.18362.0\Uno.UI.Toolkit.dll" in item list "ReferencePath" does not define a value for metadata "CopyLocal". In order to use this metadata, either qualify it by specifying %(ReferencePath.CopyLocal), or ensure that all items in this list define a value for this metadata.`

I researched something more ([Post 1](https://stackoverflow.com/questions/62468742/unable-to-build-winforms-app-due-to-referencepath-does-not-define-a-value-for), [Post 2](https://stackoverflow.com/questions/62468742/unable-to-build-winforms-app-due-to-referencepath-does-not-define-a-value-for)), but those solutions imply modify something on the compiler that is different from Uno.UI.Toolkit.dll, and I'm not sure about what to do with that

## Using console
So I tried to use the console with the command from [here](https://github.com/unoplatform/uno/discussions/8595): `dotnet publish -r win-x64 -p:PublishSingleFile=true --self-contained false` but I get this error:
`C:\Users\92lui\.nuget\packages\microsoft.windowsappsdk\1.2.221109.1\buildTransitive\Microsoft.WindowsAppSDK.SelfContain
ed.targets(42,5): error : The platform 'AnyCPU' is not supported for Self Contained mode. [PathToApp\App\App\App.Windows\App.Windows.csproj]`

First I tried to use `dotnet publish -r win-x64 -p:PublishSingleFile=true --no-self-contained` but the error was the same

Then I tried to use any other combination, even just `dotnet publish` but it was always the same error, and I don't know where that Any CPU comes from since I'm not using "Any CPU" config anywhere

## Just copying the Debug result
Finally, I tried to just copy the Debug program (the one that results when you click Debug App.Windows) and it's the same: it opens but closes immediately. It's weird but using the installed app (From the startup bar) it works perfectly, but executing the program doesn't

So **how do I have to publish an Uno app?** And how do I have to publish if I want to get programs for every platform (x86, x64 and ARM), specially if Any CPU usually doesn't work properly?

# About Android
Publishing for Android and iOS seems to work fine, although the result is an .abb file and I want an APK, I haven't researched about this because first I want to publish for Windows, but any help will be appreciated

Thanks for reading :s

**Comment by jeromelaban** [comment, marked as answer]
When building Windows AppSDK apps (the `.Windows` head), the standard ".NET publish" is not the right operation. You'll need to use the **Package and Publish** menu:

![image](https://user-images.githubusercontent.com/5839577/218177729-167d3e9b-e063-45bf-a327-a5a437e418b1.png)

Same goes for the CLI build to create and MSIX, where you're facing the fact [that the WinAppSDK does not support `dotnet build`](https://github.com/microsoft/WindowsAppSDK/issues/1652#issuecomment-981948850). You'll need to follow [this document for this](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/single-project-msix?tabs=csharp#automate-building-and-packaging-your-single-project-msix-app).

For more details, see [this documentation from Microsoft](https://learn.microsoft.com/en-us/windows/apps/package-and-deploy/self-contained-deploy/deploy-self-contained-apps).

For android, you can follow [this documentation](https://learn.microsoft.com/en-us/dotnet/maui/android/deployment/publish-cli?view=net-maui-7.0#publish) from the .NET SDK.

**Reply by LuisAlfredo92** [reply]
Thanks, it worked perfectly!

I guess this is to use it from Microsoft Store, so I'll have to do a UWP or WPF app to make it available as .exe too

Thanks for the help!

**Reply by Nurgo** [reply]
This works for a packaged Windows application, but how do you get a clean output directory for an unpackaged application, with only the files needed to launch the program? As LuisAlfredo92 mentioned, the "Publish" function doesn't work for Windows Uno Platform (WinUI) projects. You get a "Cannot resolve Assembly or Windows Metadata file" error.

**Reply by nickrandolph** [reply]
For unpackaged are you setting WindowsAppSDKSelfContained to true? If you don't, you need to work out how you're going to ensure the WinAppSdk runtime is installed on the target machine.

**Reply by Nurgo** [reply]
Yes, WindowsAppSDKSelfContained is set to true.

**Comment by liugangnhm** [comment]
> This works for a packaged Windows application, but how do you get a clean output directory for an unpackaged application, with only the files needed to launch the program? As LuisAlfredo92 mentioned, the "Publish" function doesn't work for Windows Uno Platform (WinUI) projects. You get a "Cannot resolve Assembly or Windows Metadata file" error.

same error even just publishing to directory,

**Reply by jeromelaban** [reply]
@liugangnhm would you mind opening another discussion with issues you're getting, which command line and platform you're using? Make sure to [include a binlog](https://platform.uno/docs/articles/uno-builds-troubleshooting.html).
